File: Packages/PassthroughCore/Tests/PassthroughCoreTests/ProtocolFixtureTests.swift
import XCTest
import Network
@testable import PassthroughCore

/// Checks this package against protocol/fixtures.json, the same file the
/// Android tests read, so the phone and Mac sides can't drift apart.
final class ProtocolFixtureTests: XCTestCase {
    private func fixtures() throws -> [String: Any] {
        var dir = URL(fileURLWithPath: #filePath).deletingLastPathComponent()
        while dir.path != "/" {
            let file = dir.appendingPathComponent("protocol/fixtures.json")
            if FileManager.default.fileExists(atPath: file.path) {
                return try XCTUnwrap(JSONSerialization.jsonObject(with: Data(contentsOf: file)) as? [String: Any])
            }
            dir.deleteLastPathComponent()
        }
        throw XCTSkip("protocol/fixtures.json not found")
    }

    func testConstantsMatch() throws {
        let f = try fixtures()
        XCTAssertEqual(f["protocolVersion"] as? Int, PassthroughProtocol.version)
        let ports = try XCTUnwrap(f["ports"] as? [String: Int])
        XCTAssertEqual(ports["socks"], Int(PassthroughProtocol.defaultSOCKSPort))
        XCTAssertEqual(ports["control"], Int(PassthroughProtocol.defaultControlPort))
        XCTAssertEqual(f["pairingCodeLifetimeSeconds"] as? Double, PassthroughProtocol.pairingCodeLifetime)
        let failures = try XCTUnwrap(f["pairingFailures"] as? [String])
        XCTAssertEqual(failures.compactMap(PairingFailure.init(rawValue:)).count, failures.count)
    }

    func testTokensHashAndValidateAlike() throws {
        for t in try XCTUnwrap(fixtures()["tokens"] as? [[String: Any]]) {
            let token = try XCTUnwrap(t["token"] as? String)
            XCTAssertEqual(PairingToken.hash(token), t["sha256"] as? String, token)
            XCTAssertEqual(PairingToken.isWellFormed(token), t["wellFormed"] as? Bool, token)
        }
        XCTAssertTrue(PairingToken.isWellFormed(PairingToken.generate()))
    }

    /// Every field survives a decode and re-encode, and nothing extra (like nulls) is added.
    func testMessagesRoundTrip() throws {
        for m in try XCTUnwrap(fixtures()["messages"] as? [[String: Any]]) {
            let line = try XCTUnwrap(m["line"] as? String)
            var buffer = LineBuffer()
            let framed = try buffer.append(Data((line + "\n").utf8))
            XCTAssertEqual(framed.count, 1)
            let decoded = try ControlEnvelope.decode(framed[0])
            let original = try JSONSerialization.jsonObject(with: Data(line.utf8)) as? NSDictionary
            let reencoded = try JSONSerialization.jsonObject(with: decoded.encodedLine()) as? NSDictionary
            XCTAssertEqual(original, reencoded, line)
        }
    }

    func testFormatsMatch() throws {
        let f = try fixtures()
        for r in try XCTUnwrap(f["rates"] as? [[String: Any]]) {
            let rate = ByteFormat.rate(try XCTUnwrap(r["bytesPerSecond"] as? Double))
            XCTAssertEqual(rate.value, r["value"] as? String)
            XCTAssertEqual(rate.unit, r["unit"] as? String)
        }
        for d in try XCTUnwrap(f["durations"] as? [[String: Any]]) {
            XCTAssertEqual(ByteFormat.duration(try XCTUnwrap(d["seconds"] as? Double)), d["text"] as? String)
        }
    }

    func testDNSRedirectRules() throws {
        let f = try XCTUnwrap(fixtures()["dnsRedirect"] as? [String: Any])
        XCTAssertEqual(f["port"] as? Int, Int(DNSRedirect.port))
        XCTAssertEqual(f["fallback"] as? String, DNSRedirect.fallback)
        for d in try XCTUnwrap(f["destinations"] as? [[String: Any]]) {
            let text = try XCTUnwrap(d["address"] as? String)
            let port = UInt16(try XCTUnwrap(d["port"] as? Int))
            let portBytes = [UInt8(port >> 8), UInt8(port & 0xFF)]
            let raw: Data
            if let v4 = IPv4Address(text) { raw = Data([SOCKS5.AddressType.ipv4]) + v4.rawValue + portBytes }
            else { raw = Data([SOCKS5.AddressType.ipv6]) + (try XCTUnwrap(IPv6Address(text))).rawValue + portBytes }
            let address = try XCTUnwrap(SOCKS5.Address(raw: raw))
            XCTAssertEqual(DNSRedirect.applies(to: address), d["redirect"] as? Bool, "\(text):\(port)")
        }
        for c in try XCTUnwrap(f["serverChoice"] as? [[String: Any]]) {
            XCTAssertEqual(DNSRedirect.server(from: try XCTUnwrap(c["system"] as? [String])), c["chosen"] as? String)
        }
        // Reads the real resolver configuration without crashing; the list depends on the machine.
        _ = DNSRedirect.systemServers()
    }
}

File: android/app/src/test/java/dev/dpatel/passthrough/core/ProtocolFixtureTests.kt
package dev.dpatel.passthrough.core

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.double
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long
import org.junit.Assert.assertEquals
import org.junit.Test
import java.io.File
import java.io.StringReader

/**
 * Checks this app against protocol/fixtures.json, the same file the Swift
 * tests read, so the phone and Mac sides can't drift apart.
 */
class ProtocolFixtureTests {
    private val fixtures: JsonObject by lazy {
        var dir: File? = File("").absoluteFile
        while (dir != null && !File(dir, "protocol/fixtures.json").exists()) dir = dir.parentFile
        val file = File(requireNotNull(dir) { "protocol/fixtures.json not found above ${File("").absolutePath}" }, "protocol/fixtures.json")
        Json.parseToJsonElement(file.readText()).jsonObject
    }

    @Test fun constantsMatch() {
        assertEquals(fixtures["protocolVersion"]!!.jsonPrimitive.int, PassthroughProtocol.VERSION)
        val ports = fixtures["ports"]!!.jsonObject
        assertEquals(ports["socks"]!!.jsonPrimitive.int, PassthroughProtocol.DEFAULT_SOCKS_PORT)
        assertEquals(ports["control"]!!.jsonPrimitive.int, PassthroughProtocol.DEFAULT_CONTROL_PORT)
        assertEquals(fixtures["pairingCodeLifetimeSeconds"]!!.jsonPrimitive.long * 1000, PassthroughProtocol.PAIRING_CODE_LIFETIME_MS)
        assertEquals(fixtures["pairingFailures"]!!.jsonArray.map { it.jsonPrimitive.content }, PairingFailure.entries.map { it.wire })
    }

    @Test fun tokensHashAndValidateAlike() {
        for (t in fixtures["tokens"]!!.jsonArray.map { it.jsonObject }) {
            val token = t["token"]!!.jsonPrimitive.content
            assertEquals(token, t["sha256"]!!.jsonPrimitive.content, PairingToken.hash(token))
            assertEquals(token, t["wellFormed"]!!.jsonPrimitive.boolean, PairingToken.isWellFormed(token))
        }
    }

    /** Every field survives a decode and re-encode, and nothing extra (like nulls) is added. */
    @Test fun messagesRoundTrip() {
        for (m in fixtures["messages"]!!.jsonArray) {
            val line = m.jsonObject["line"]!!.jsonPrimitive.content
            val decoded = LineReader(line.plus("\n").byteInputStream()).next()!!.let(ControlEnvelope::decode)
            val encoded = String(decoded.encodeLine()).trimEnd('\n')
            assertEquals(line, normalize(Json.parseToJsonElement(line)), normalize(Json.parseToJsonElement(encoded)))
        }
    }

    @Test fun formatsMatch() {
        for (r in fixtures["rates"]!!.jsonArray.map { it.jsonObject }) {
            val expected = r["value"]!!.jsonPrimitive.content to r["unit"]!!.jsonPrimitive.content
            assertEquals(expected, ByteFormat.rate(r["bytesPerSecond"]!!.jsonPrimitive.double))
        }
        for (d in fixtures["durations"]!!.jsonArray.map { it.jsonObject }) {
            assertEquals(d["text"]!!.jsonPrimitive.content, ByteFormat.duration(d["seconds"]!!.jsonPrimitive.long))
        }
    }

    @Test fun lineReaderSkipsBlankLinesAndCapsLength() {
        val reader = LineReader(StringReader("a\n\nb\n").readText().byteInputStream())
        assertEquals("a", reader.next())
        assertEquals("b", reader.next())
        assertEquals(null, reader.next())
        val long = LineReader("123456789".byteInputStream(), limit = 8)
        assertEquals(true, runCatching { long.next() }.isFailure)
    }

    @Test fun dnsRedirectRules() {
        val f = fixtures["dnsRedirect"]!!.jsonObject
        assertEquals(f["port"]!!.jsonPrimitive.int, DnsRedirect.PORT)
        assertEquals(f["fallback"]!!.jsonPrimitive.content, DnsRedirect.FALLBACK)
        for (d in f["destinations"]!!.jsonArray.map { it.jsonObject }) {
            val ip = java.net.InetAddress.getByName(d["address"]!!.jsonPrimitive.content)
            val address = Socks5.Address.parse(Socks5.rawAddress(ip, d["port"]!!.jsonPrimitive.int))!!
            assertEquals(address.toString(), d["redirect"]!!.jsonPrimitive.boolean, DnsRedirect.applies(address))
        }
        for (c in f["serverChoice"]!!.jsonArray.map { it.jsonObject }) {
            val system = c["system"]!!.jsonArray.map { java.net.InetAddress.getByName(it.jsonPrimitive.content) }
            assertEquals(java.net.InetAddress.getByName(c["chosen"]!!.jsonPrimitive.content), DnsRedirect.server(system))
        }
    }

    /** Numbers compare by value (1.0 == 1, 1.7580624E9 == 1758062400.0); strings and booleans by content. */
    private fun normalize(e: JsonElement): Any? = when (e) {
        is JsonObject -> e.mapValues { normalize(it.value) }
        is JsonArray -> e.map { normalize(it) }
        is JsonPrimitive -> if (e.isString) e.content else e.content.toBooleanStrictOrNull() ?: e.content.toDoubleOrNull()
    }
}
